---
title: "5GPM"
author: "Sipko"
date: "October 18, 2016"
output:
  html_document:
    fig_width: 14
    fig_height: 7
    max-width: 1600px;
    width: 1600px;
    out.width: 1600px;
---



```{r setup, include=FALSE}
#knitr::opts_chunk$set(fig.width=12, fig.height=8, fig.path='Figs/',echo=FALSE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(echo = FALSE, warning=FALSE, message=FALSE)
```

```{r set-options, echo=FALSE, cache=FALSE}
options(width = 1600)
```

```{r libraries, include=FALSE, echo=FALSE, warning=FALSE, message=FALSE}
#install.packages('devtools') #uncomment if libraries are missing
if(!require(devtools)) {
install.packages("devtools"); require(devtools)} #load / install+load devtools
library(devtools) 
if(!require(scales)) {
install.packages("scales"); require(scales)} #load / install+load devtools
library(scales)
if(!require(plyr)) {
install.packages("plyr"); require(plyr)} #load / install+load devtools
library(plyr)
if(!require(stringr)) {
install.packages("stringr"); require(stringr)} #load / install+load devtools
library(stringr)
if(!require(data.table)) {
install.packages("data.table"); require(data.table)} #load / install+load devtools
library(data.table)
if(!require(ggplot2)) {
install.packages("ggplot2"); require(ggplot2)} #load / install+load devtools
library(ggplot2)
#devtools::install_github("RcppCore/Rcpp",dep=TRUE) #uncomment if libraries are missing
#devtools::install_github("slowkow/ggrepel",dep=TRUE) #uncomment if libraries are missing
library(ggrepel)
if(!require(gridExtra)) {
install.packages("gridExtra"); require(gridExtra)} #load / install+load devtools
library(gridExtra)

#functions
PCAmetadata <- function(GPM, GPMnames,title)
{
  gpm_eigen=prcomp(GPM, center=TRUE,Scale=TRUE)
  print("Proportion of Variance explained: ")
  print(summary(gpm_eigen)$importance["Proportion of Variance",1:10])
  plot(gpm_eigen, main=title, type="l")
  pcInfo <- cbind(gpm_eigen$rotation,sampleIDs=noquote(GPMnames))
  #colnames(pcInfo)[length(pcInfo[1,])]="sampleIDs"
  pcInfo[,"sampleIDs"]=noquote(pcInfo[,"sampleIDs"])
  colindex=match(pcInfo[,"sampleIDs"],meta[,"externalSampleID"])
  pcInfo <- cbind.data.frame(pcInfo,Cell.type=meta[colindex,"Cell.type"])
  pcInfo <- cbind.data.frame(pcInfo,Study.group=meta[colindex,"Study.group"])
  pcInfo <- cbind.data.frame(pcInfo,Causal.gene=meta[colindex,"Causal.gene"])
  pcInfo <- cbind.data.frame(pcInfo,Familie.nummer=meta[colindex,"Familie.nummer"])
  
  #not sure if this is needed but for some reason values are not considered numeric otherwise
  for(i in c(1:100))
  {
    pcInfo[,i]=as.numeric(as.character((pcInfo[,i])))
  }
  return(pcInfo)
}
getGenes <- function(GPM,GPMnames,addMeta=TRUE)
{
  genes=c("ENSG00000012048","ENSG00000139618","ENSG00000145191","ENSG00000127554","ENSG00000170266","ENSG00000198910","ENSG00000134571","ENSG00000092054","ENSG00000148400","ENSG00000165917")

  subsetGPM=GPM[genes,]
  conversion=data.frame(fread("E:/Groningen/Data/Annotation/GRCh38/DiseaseGenes/EnsgToSymbol.txt"),row.names=1)
  rownames(subsetGPM)=conversion[rownames(subsetGPM),1]
  subsetGPMtrans=t(subsetGPM)
  if(addMeta)
  {
    colindex=match(colnames(GPMnames)[-1],meta[,"externalSampleID"])
  
    subsetGPMtrans=cbind.data.frame(subsetGPMtrans,causal.gene=meta[colindex,"Causal.gene"])
    subsetGPMtrans=cbind.data.frame(subsetGPMtrans,Cell.type=meta[colindex,"Cell.type"])
  }
  return(subsetGPMtrans)
}

plotPairExpression <- function(gene, subsetGPMtrans_N, subsetGPMtrans_C)
{
    #PLOT BASED ON NORMAL EXPRESSION
  CellType=subsetGPMtrans_N[,"Cell.type"]
  labels1=replace(subsetGPMtrans_N[,"causal.gene"],subsetGPMtrans_N[,"causal.gene"]!=gene,"")
  p1 <- ggplot(subsetGPMtrans_N, aes(x=c(1:length(subsetGPMtrans_N[,1])),y=subsetGPMtrans_N[,gene], label=labels1,color=CellType))
  plot1=p1 + geom_point(size=1) + geom_text_repel()+ggtitle(paste("Uncorrected",gene)) + xlab("PC1") + ylab("PC2")
  
  #get the genes from the matrixes
  Cell_Type=subsetGPMtrans_C[,"Cell.type"]
  labels2=replace(subsetGPMtrans_C[,"causal.gene"],subsetGPMtrans_C[,"causal.gene"]!=gene,"")
  p2 <- ggplot(subsetGPMtrans_C, aes(x=c(1:length(subsetGPMtrans_C[,1])),y=subsetGPMtrans_C[,gene], label=labels2,color=Cell_Type))
  plot2=p2 + geom_point() + geom_text_repel()+ggtitle(paste("Corrected",gene)) + xlab("PC1") + ylab("PC2")
  print(plot1)
  print(plot2)
  #grid.arrange(plot1,plot2, ncol=2)
}

printOutlierInfo <- function(sampleToGene,GPM_extra_Z_NotCor)
{
  for(i in 1:length(rownames(sampleToGene)))
  {
    sample=rownames(sampleToGene)[i]
    geneVal=GPM_extra_Z_NotCor[as.character(sampleToGene[i,"Ensg.gene"]),sample]
    larger=length(GPM_extra_Z_NotCor[GPM_extra_Z_NotCor[,sample]>=geneVal,sample])
    smaller=length(GPM_extra_Z_NotCor[GPM_extra_Z_NotCor[,sample]<=geneVal,sample])
    largerThan20=length(GPM_extra_Z_NotCor[GPM_extra_Z_NotCor[,sample]>=20,sample])
    
    mindistance=smaller
    if(mindistance>larger)
      mindistance=larger
    
    cat("\nSample:",sample)
    cat("\tGene:",as.character(sampleToGene[i,"Causal.gene"]))
    cat("\tlarger:", larger)
    cat("\tsmaller:", smaller)
    cat("\tMin_distance:", mindistance)
    cat("\tNum_genes >20:", largerThan20)
  }
}

loadDown <- function(filename, chromInfo)
{
  DownNames_N <- fread(filename)
  Down_N <- data.frame(DownNames_N, row.names=1);
  index_N=match(rownames(Down_N),rownames(chromInfo))
  Down_N_extra=cbind.data.frame(Down_N, Color=chromInfo[index_N,"Color"])
}

plotPerSample <- function(sampleToGene, GPM_extra_Z_NotCor, GPM_extra_C_Z)
{
  for(sample in rownames(sampleToGene))
  {
    print(sample)
    
    geneVal_U=GPM_extra_Z_NotCor[as.character(sampleToGene[sample,"Ensg.gene"]),sample]
    larger_U=length(GPM_extra_Z_NotCor[GPM_extra_Z_NotCor[,sample]>=geneVal_U,sample])
    smaller_U=length(GPM_extra_Z_NotCor[GPM_extra_Z_NotCor[,sample]<=geneVal_U,sample])
    largerThan20_U=length(GPM_extra_Z_NotCor[GPM_extra_Z_NotCor[,sample]>=20,sample])
    
    gene=as.character(sampleToGene[sample,"Causal.gene"])
    
    labels1=replace(GPM_extra_Z_NotCor[,"Causal.gene"],GPM_extra_Z_NotCor[,"Causal.gene"]!=gene,"")
    p1 <- ggplot(GPM_extra_Z_NotCor, aes(x=c(1:length(GPM_extra_Z_NotCor[,1])),y=GPM_extra_Z_NotCor[,sample], label=labels1))
    plot1=p1 + geom_point(size=0.2) + geom_text_repel(size=3,colour="red")+ ylim(-20, 20)+ ggtitle(paste("Uncorrected", sample," ",gene," Larger:",larger_U," Smaller:", smaller_U, " N_largerThan20:", largerThan20_U)) + xlab("Genes") + ylab("signal")
    
    geneVal_C=GPM_extra_C_Z[as.character(sampleToGene[sample,"Ensg.gene"]),sample]
    larger_C=length(GPM_extra_C_Z[GPM_extra_C_Z[,sample]>geneVal_C,sample])
    smaller_C=length(GPM_extra_C_Z[GPM_extra_C_Z[,sample]<geneVal_C,sample])
    largerThan20_C=length(GPM_extra_C_Z[GPM_extra_C_Z[,sample]>20,sample])
    
    labels2=replace(GPM_extra_C_Z[,"Causal.gene"],GPM_extra_C_Z[,"Causal.gene"]!=gene,"")
    p2 <- ggplot(GPM_extra_C_Z, aes(x=c(1:length(GPM_extra_C_Z[,1])),y=GPM_extra_C_Z[,sample], label=labels2))
    plot2=p2 + geom_point(size=0.2) + geom_text_repel(size=3,colour="red")+ ylim(-20, 20)+ ggtitle(paste("Corrected", sample," ",gene," Larger:",larger_C," Smaller:", smaller_C, " N_largerThan20:", largerThan20_C)) + xlab("Genes") + ylab("signal")
    grid.arrange(plot1,plot2,nrow=2)
  }
}

quadPlotDown <- function(Down_N_extra,Down_C_extra,Down_C_Stdev_extra,Down_C_Z_extra, start, end)
{
  for(s in c(start:end))
  {
    print(colnames(Down_N_extra)[s],size="\\fontsize{20pt}{10pt}\\selectfont")
    p1 <- ggplot(Down_N_extra, aes(x=c(1:length(Down_N_extra[,1])),y=Down_N_extra[,s]))
        plot1=p1 + geom_point(size=0.2, alpha=0.4, colour=Down_N_extra[,"Color"]) + ggtitle(paste("Uncorrected"))+ylim(-3,3)+ xlab("Genes") + ylab("signal")
        
    p2 <- ggplot(Down_C_extra, aes(x=c(1:length(Down_C_extra[,1])),y=Down_C_extra[,s]))
        plot2=p2 + geom_point(size=0.2, alpha=0.4,colour=Down_C_extra[,"Color"]) + ggtitle(paste("Corrected"))+ylim(-3,3)+ xlab("Genes") + ylab("signal")
        
    p3 <- ggplot(Down_C_Stdev_extra, aes(x=c(1:length(Down_C_Stdev_extra[,1])),y=Down_C_Stdev_extra[,s]))
        plot3=p3 + geom_point(size=0.2, alpha=0.4,colour=Down_C_Stdev_extra[,"Color"]) + ggtitle(paste("Corrected/(stdev_genes_uncorrected_healthy)"))+ xlab("Genes") + ylab("signal")
        
    p4 <- ggplot(Down_C_Z_extra, aes(x=c(1:length(Down_C_Z_extra[,1])),y=Down_C_Z_extra[,s]))
        plot4=p4 + geom_point(size=0.2, alpha=0.4,colour=Down_C_Z_extra[,"Color"]) + ggtitle(paste("Z-score genes corrected (based on avg&stdev genes corrected BBMRI)"))+ylim(-10,10)+ xlab("Genes") + ylab("signal")
        
        grid.arrange(plot1,plot2,plot3,plot4,nrow=4)
  }
}

```
## Down syndrome samples
First we tested if the outlier expression is detected in the down-syndrome samples in data that is corrected for the first 230 principal components obtained from the public "healthy" 13600 samples (there are 230 PCs cronbach >0.7).

For each down syndrome sample we made 4 plots:
1. Uncorrected
2. Corrected 230 PCs
3. Corrected 230 PCs and divided by the standard deviation observed in the uncorrected 13600 healhty samples (for each gene)
4. Z-scores of the Corrected for 230 PCs. The z-scores are calculated using the averages and standard deviations observed in the BBMRI samples also corrected for the first 230 "healthy" PCs.

```{r Down_syndrome, echo = FALSE, warning=FALSE, message=FALSE,fig_width=15,fig_height=7}
wd="E:/Groningen/Data/Juha/Genes31995/Healthy/PCA/31.07.pc1.illumina.genes.expressed_DownSamples/"
#not corrected data
chromInfo=data.frame(fread("E:/Groningen/Data/Annotation/GRCh38/ChromosomeInfo/GenePositionInfo.txt"),row.names=1);

#not corrected
Down_N_extra=loadDown(paste(wd,"PC_1-0_12000highest.txt",sep=""),chromInfo)
#corrected
Down_C_extra=loadDown(paste(wd,"PC_1-1-230_12000highest.txt",sep=""),chromInfo)
#Corrected divided by stdev of healthy stamples (uncorrected)
Down_C_Stdev_extra=loadDown(paste(wd,"PC_1-1-230_DevidedBySTdevs_12000highest.txt",sep=""),chromInfo)
#Corrected z-scores based on avg and stdev from BBMRI (corrected 230 PCs)
Down_C_Z_extra=loadDown(paste(wd,"PC_1-1-230_zScores_12000highest.txt",sep=""),chromInfo)

quadPlotDown(Down_N_extra,Down_C_extra,Down_C_Stdev_extra,Down_C_Z_extra,3,19);
```

## PCA on samples
We compared the 5GPM samples based on the first 2 principal components for the corrected an uncorrected data. Comparsion of variance explained of the first 10 PCs before and after correcting for the Principal componenets identified from the public data.

```{r PCA_5GPMsamples, echo = FALSE, warning=FALSE, message=FALSE,results='hide'}
###load uncorrected data
par(mfrow = c(1, 2))
# par(cex = 0.6)
# par(mar = c(3, 3, 0, 0), oma = c(1, 1, 1, 1))
wd="E:/Groningen/Data/RNAseq_clinic/5GPM/GRCh38/"
GPMnames_N <- fread(paste(wd,"counts_GENES_ScaffoldGenesRemoved_DiscardedRemoved.DESeqNorm.Log2_RNAids.txt",sep=""))
GPM_N <- data.frame(GPMnames_N, row.names=1);
meta <- read.table("E:/Groningen/Data/RNAseq_clinic/RNAseq_Info.txt", sep="\t", header=TRUE)
rownames(meta)=make.names(meta[,"externalSampleID"], unique=TRUE)
pcInfo_N=PCAmetadata(GPM_N, colnames(GPMnames_N)[-1],"uncorrected")

###load corrected data
wd="E:/Groningen/Data/Juha/Genes31995/Healthy/PCA/counts_GENES_ScaffoldGenesRemoved_DiscardedRemoved_RNAids/"
GPMnames_C <- fread(paste(wd,"PC_1-1-230_.txt",sep=""))
GPM_C <- data.frame(GPMnames_C, row.names=1);
rownames(meta)=make.names(meta[,"externalSampleID"], unique=TRUE)
pcInfo_C=PCAmetadata(GPM_C[,-(1:2)], colnames(GPMnames_C)[-(1:3)],"corrected")
```

```{r plotPCA, echo = FALSE,  warning=FALSE, message=FALSE}
#plot the uncorrected data
Cell_Type_N=pcInfo_N[,"Cell.type"]
labels=pcInfo_N[,"Causal.gene"][which(pcInfo_N[,"Causal.gene"]=="NOTCH1")]
p1<- ggplot(pcInfo_N, aes(x=pcInfo_N[,1],y=pcInfo_N[,2], label=pcInfo_N[,"Causal.gene"], color=Cell_Type_N))
plot1=p1 + geom_point() + geom_label_repel() + xlab("PC1") + ylab("PC2") + ggtitle("uncorrected")
#Plot the corrected data
Cell_Type_C=pcInfo_C[,"Cell.type"]
p2<- ggplot(pcInfo_C, aes(x=pcInfo_C[,1],y=pcInfo_C[,2], label=pcInfo_C[,"Causal.gene"], color=Cell_Type_C))
plot2=p2 + geom_point() + geom_label_repel() + xlab("PC1") + ylab("PC2") + ggtitle("corrected")
print(plot1)
print(plot2)
#grid.arrange(plot1, plot2, ncol=2)
```

##Outlier expression per gene 5GPM
Next we plot the expression of each gene (both corrected and not corrected) based on the 5GPM samples only.

One sample is grey, as it is not in the sample sheet, but it is a sample in the matrix file.The two NOTCH1 outlier samples originate from 1 sample and are sequenced in duplicate. Both of which have 69% reads mapping (although not exact same number).

```{r boxPlotsGenes, echo = FALSE, warning=FALSE, message=FALSE}
subsetGPMtrans_N=getGenes(GPM_N, GPMnames_N)
subsetGPMtrans_C=getGenes(GPM_C,GPMnames_C) 
conversion=data.frame(fread("E:/Groningen/Data/Annotation/GRCh38/DiseaseGenes/EnsgToSymbol.txt"),row.names=1)
genes=conversion[,1]

###NOTCH1, EIF2B5, GLB1, BRCA1,BRCA2,L1CAM,MYBPC3,RAPSN,MYH7
#BRCA2, NOTCH1, GLB1, L1CAM, BRCA1,GFER
#EIF2B5
for(gene in genes)
{
  plotPairExpression(gene, subsetGPMtrans_N, subsetGPMtrans_C)
}
```

##Outlier expression per gene 5GPM compared to BBMRI
Next we plot the expression of each gene ( corrected and not corrected) for 5GPM and BBMRI samples
```{r boxPlotsGenesBBMRI, echo = FALSE, warning=FALSE, message=FALSE}
###now do the same with the BBMRI samples merged
#BBMRI normal
BBMRInames_N <- fread("E:/Groningen/Data/bbmri/GCRh38/DESeqNorm/counts_GENES_ScaffoldGenesRemoved.DESeqNorm.Log2.txt")
BBMRI_N <- data.frame(BBMRInames_N, row.names=1);
#pcInfoBBMRI_N=PCAmetadata(BBMRI_N, colnames(GPMnames_N)[-1])
subsetBBMRItrans_N=getGenes(BBMRI_N,GPMnames_N,FALSE) 
subsetBBMRItrans_N=data.frame(subsetBBMRItrans_N);
colIndex_N=match(colnames(subsetGPMtrans_N),colnames(subsetBBMRItrans_N))
merge_N=rbind.fill(subsetGPMtrans_N, subsetBBMRItrans_N[,])

##BBMRI corrected
BBMRInames_C <- fread("E:/Groningen/Data/Juha/Genes31995/Healthy/PCA/BBMRI/PC_1-1-230_.txt")
BBMRI_C <- data.frame(BBMRInames_C, row.names=1);
#pcInfoBBMRI_N=PCAmetadata(BBMRI_C, colnames(GPMnames_C)[-(1:3)])
subsetBBMRItrans_C=getGenes(BBMRI_C,GPMnames_C,FALSE) 
colIndex_C=match(colnames(subsetGPMtrans_C),colnames(subsetBBMRItrans_C))
merge_C=rbind.fill(data.frame(subsetGPMtrans_C), data.frame(subsetBBMRItrans_C[,]))

for(gene in genes)
{
  plotPairExpression(gene, merge_N, merge_C)
}

#p <- ggplot(merge_N, aes(x=c(1:length(merge_N[,1])),y=merge_N[,"NOTCH1"], label=merge_N[,"causal.gene"],color=merge_N[,"Cell.type"]))
#p + geom_point(size=1) + geom_text_repel(size=2)

#p <- ggplot(merge_C, aes(x=c(1:length(merge_C[,1])),y=merge_C[,"NOTCH1"], label=merge_C[,"causal.gene"],color=merge_C[,"Cell.type"]))
#p + geom_point(size=1) + geom_text_repel(size=2)

#p <- ggplot(subsetGPMtrans_N, aes(x=c(1:length(subsetGPMtrans_N[,1])),y=subsetGPMtrans_N[,"NOTCH1"], label=subsetGPMtrans_N[,"causal.gene"],color=subsetGPMtrans_N[,"Cell.type"]))
#p + geom_point(size=1) + geom_text_repel(size=2)

#p <- ggplot(subsetGPMtrans_C, aes(x=c(1:length(subsetGPMtrans_C[,1])),y=subsetGPMtrans_C[,"NOTCH1"], label=subsetGPMtrans_C[,"causal.gene"],color=subsetGPMtrans_C[,"Cell.type"]))
#p + geom_point(size=1) + geom_text_repel(size=2)
```


##Plotting outlier expression per sample instead of per gene based on z-scores based on averages and stdevs per gene in BBMRI data
As before we plot the scores per sample. The z-scores per gene are calculated based on the averages and standard deviation observed in the BBMRI data.

We also report the number of genes that have a z-score larger than 20 for each sample as we feel this is an idication of how noisy the data is.

```{r zscores, echo = FALSE,fig_width=15,fig_height=15}
#NOTCH1: RNA_001011,RNA_001011.1,RNA_000979
#GLB1: RNA.001128
#L1CAM: RNA_000811,RNA_000947
#EIF2B5: GPM_extra_Z["ENSG00000145191","RNA.001110"]
#BRCA2:  GPM_extra_Z["ENSG00000139618","RNA_000751"]
#BRCA1: GPM_extra_Z["ENSG00000012048","RNA_000939"]
#MYBPC3: "RNA_00958", "RNA_00958.1"
#RAPSN: "RNA.001126", "RNA.001126.1" 
#GFER: RNA.001103
#MYH7: c("RNA_001073","RNA_001073.1","RNA_001075","RNA_001075.1")
wd="E:/Groningen/Data/Juha/Genes31995/Healthy/PCA/counts_GENES_ScaffoldGenesRemoved_DiscardedRemoved_RNAids/"
#load uncorrected z-scores
GPMnames_N_Z <- fread(paste(wd,"PC_1-0__zScores.txt",sep=""))
GPMnames_N_Z <- data.frame(GPMnames_N_Z, row.names=1);

index=match(rownames(GPMnames_N_Z),rownames(conversion))
GPM_extra_Z_NotCor=cbind.data.frame(GPMnames_N_Z, Causal.gene=conversion[index,1])
#load corrected z-scores
GPMnames_C_Z <- fread(paste(wd,"PC_1-1-230__zScores.txt",sep=""))
GPM_C_Z <- data.frame(GPMnames_C_Z, row.names=1);
sampleToGene=read.table("E:/Groningen/Data/RNAseq_clinic/SampleToGene.txt", sep="\t",row.names=1, header=TRUE)
index=match(rownames(GPM_C_Z),rownames(conversion))
GPM_extra_C_Z=cbind.data.frame(GPM_C_Z, Causal.gene=conversion[index,1])
plotPerSample(sampleToGene, GPM_extra_Z_NotCor, GPM_extra_C_Z)
```

##Summary

This table summarizes the number of genes that have a higher z-score per sample and how many lower. "Min_distance" indicates how close it is to being the highest or the lowest expressed gene (whichever number is lower),closer to 0 is better

`````{r zscoresTable, echo = FALSE}
print("Not corrected z-score")
printOutlierInfo(sampleToGene,GPM_extra_Z_NotCor)
print("Corrected z-score")
printOutlierInfo(sampleToGene,GPM_extra_C_Z)
```

